### ISSUE TITLE
Interoperation of UniStr and String

### ISSUE THREAD

**Issue by bkamins** [JuliaString/Strs.jl#57]
A small issue when mixing `UniStr` and `String`:

```
julia> y = [Strs.UniStr("ąęł")]
1-element Array{_UCS2Str,1}:
 "ąęł"

julia> x = ["ael"]
1-element Array{String,1}:
 "ael"

julia> [x y]
ERROR: TypeError: in copyto!, in typeassert, expected String, got UTF8Str
Stacktrace:
 [1] setindex! at .\array.jl:688 [inlined]
 [2] copyto!(::Array{String,2}, ::Int64, ::Array{_UCS2Str,1}, ::Int64, ::Int64) at .\abstractarray.jl:729
 [3] typed_hcat(::Type{String}, ::Array{String,1}, ::Array{_UCS2Str,1}) at .\abstractarray.jl:1162
 [4] hcat(::Array{String,1}, ::Array{_UCS2Str,1}) at \sparsevector.jl:1046
 [5] top-level scope
```

**Comment by bkamins** [comment]
Actually you get the same problem even within Strs.jl ecosystem:

```
julia> x = [Strs.UniStr("ael")]
1-element Array{ASCIIStr,1}:
 "ael"

julia> y = [Strs.UniStr("ąęł")]
1-element Array{_UCS2Str,1}:
 "ąęł"

julia> [x y]
ERROR: TypeError: in copyto!, in typeassert, expected String, got UTF8Str
Stacktrace:
 [1] setindex! at .\array.jl:688 [inlined]
 [2] copyto!(::Array{String,2}, ::Int64, ::Array{_UCS2Str,1}, ::Int64, ::Int64) at .\abstractarray.jl:729
 [3] typed_hcat(::Type{String}, ::Array{ASCIIStr,1}, ::Array{_UCS2Str,1}) at .\abstractarray.jl:1162
 [4] hcat(::Array{ASCIIStr,1}, ::Array{_UCS2Str,1}) at \sparsevector.jl:1046
 [5] top-level scope
```

**Comment by bkamins** [comment]
The offending line is:
```
promote_rule(::Type{<:AbstractString}, ::Type{<:AbstractString}) = String
```
from *strings/basic.jl* in Base - I guess we need custom promotion rules as noted earlier to cover all standard operations:

```
Base.promote_rule(::Type{<:Strs.UniStr}, ::Type{<:Strs.UniStr}) = Strs.UniStr
Base.promote_rule(::Type{<:Strs.UniStr}, ::Type{Missing}) = Union{Strs.UniStr, Missing}
Base.promote_rule(::Type{Missing}, ::Type{<:Strs.UniStr}) = Union{Strs.UniStr, Missing}
Base.typejoin(::Type{<:Strs.UniStr}, ::Type{<:Strs.UniStr}) = Strs.UniStr
Base.typejoin(::Type{<:Strs.UniStr}, ::Type{Missing}) = Union{Strs.UniStr, Missing}
Base.typejoin(::Type{Missing}, ::Type{<:Strs.UniStr}) = Union{Strs.UniStr, Missing}
Base.promote_typejoin(::Type{<:Strs.UniStr}, ::Type{<:Strs.UniStr}) = Strs.UniStr
Base.promote_typejoin(::Type{<:Strs.UniStr}, ::Type{Missing}) = Union{Strs.UniStr, Missing}
Base.promote_typejoin(::Type{Missing}, ::Type{<:Strs.UniStr}) = Union{Strs.UniStr, Missing}
```

This should probably be checked if all this is needed (probably yes) and if this is all that is needed.

**Comment by ScottPJones** [comment]
Fixed by #81
